piccolo_api/crud/endpoints.py

```diff
@@ -1038,37 +1053,69 @@ async def patch_single(
 
         cls = self.table
 
-        try:
-            values = {
-                getattr(cls, key): getattr(model, key) for key in data.keys()
-            }
-        except AttributeError:
-            unrecognised_keys = set(data.keys()) - set(model.dict().keys())
-            return Response(
-                f"Unrecognised keys - {unrecognised_keys}.", status_code=400
+        if issubclass(cls, BaseUser):
+            old_password = (
+                await cls.select(cls.password)
+                .where(cls.email == cleaned_data["email"])
+                .first()
+                .run()
             )
 
-        if self._hook_map:
-            values = await execute_patch_hooks(
-                hooks=self._hook_map,
-                hook_type=HookType.pre_patch,
-                row_id=row_id,
-                values=values,
-            )
+            try:
+                cls._validate_password(password=cleaned_data["password"])
+                cleaned_data["password"] = cls.hash_password(
+                    cleaned_data["password"]
+                )
+            except ValueError:
+                cleaned_data["password"] = old_password["password"]
 
-        try:
+            model = self.pydantic_model_output(**cleaned_data)
+            values = {
+                getattr(cls, key): getattr(model, key)
+                for key in cleaned_data.keys()
+            }
             await cls.update(values).where(
-                cls._meta.primary_key == row_id
+                cls.email == cleaned_data["email"]
             ).run()
-            new_row = (
-                await cls.select(exclude_secrets=self.exclude_secrets)
-                .where(cls._meta.primary_key == row_id)
-                .first()
-                .run()
-            )
-            return CustomJSONResponse(self.pydantic_model(**new_row).json())
-        except ValueError:
-            return Response("Unable to save the resource.", status_code=500)
+            return Response(status_code=200)
```

Comment:
I see what you're doing. I need to think about this a bit.

Maybe this simpler logic works:

```python
values = model.dict()
if model.password:
    cls._validate_password(model.password)
    values['password'] = cls.hash_password(model.password)
else:
    values.pop('password')

await cls.update(**values).where(
    cls.email == cleaned_data["email"]
).run()
return Response(status_code=200)
```

Reply:
Unfortunately I failed to implement this. I know this is a little complicated. I’ll push another commit with exceptions and if you come up with something easier to follow, it will be great.

Reply:
I managed to change the code based on your suggestion in the last commit. I think it's much better now.